File: framework/scala/src/main/scala/soda/scala/web/http/MultipartFormData.scala
package soda.scala.web.http

import soda.scala.web.exception.ParameterMissingException

class MultipartFormData(private val partData: Map[String, List[Part]]) {

  def firstValueOpt(key: String): Option[String] = {
    partData.getOrElse(key, List.empty).headOption.map(_.bodyString)
  }

  def firstValue(key: String): String = {
    firstValueOpt(key) match {
      case Some(v) => v
      case None => throw new ParameterMissingException(key)
    }
  }

  def values(key: String): List[String] = {
    partData.getOrElse(key, List.empty).map(_.bodyString)
  }

  def firstFileOpt(key: String): Option[Array[Byte]] = {
    partData.getOrElse(key, List.empty).headOption.map(_.payload)
  }

  def firstFile(key: String): Array[Byte] = {
    firstFileOpt(key) match {
      case Some(v) => v
      case None => throw new ParameterMissingException(key)
    }
  }

  def files(key: String): List[Array[Byte]] = {
    partData.getOrElse(key, List.empty).map(_.payload)
  }

}

File: framework/kotlin/src/main/kotlin/soda/kotlin/web/http/MultipartFormData.kt
package soda.kotlin.web.http

import soda.kotlin.web.exception.ParameterMissingException

class MultipartFormData(private val partData: Map<String, List<Part>>) {

    fun firstValueOpt(key: String): String? {
        return elementOpt(key) { it.bodyString() }
    }

    fun firstValue(key: String): String {
        return firstElement(key) { it.bodyString() }
    }

    fun values(key: String): List<String> {
        return elements(key) { it.bodyString() }
    }

    fun firstFileOpt(key: String): ByteArray? {
        return elementOpt(key) { it.payload }
    }

    fun firstFile(key: String): ByteArray {
        return firstElement(key) { it.payload }
    }

    fun files(key: String): List<ByteArray> {
        return elements(key) { it.payload }
    }

    private fun <T> elements(key: String, mapper: (Part) -> T): List<T> {
        return partData.getOrElse(key) { listOf() }.map(mapper)
    }

    private fun <T> elementOpt(key: String, mapper: (Part) -> T): T? {
        return elements(key, mapper).firstOrNull()
    }

    private fun <T> firstElement(key: String, mapper: (Part) -> T): T {
        return elementOpt(key, mapper) ?: throw ParameterMissingException(key)
    }

}